### DISCUSSION TITLE
Multi-Server Mode

### DISCUSSION THREAD

**Discussion by mhils** [mitmproxy/mitmproxy#5288] [Ideas]
**TL;DR: Mitmproxy needs to be able to spawn multiple proxy servers simultaneously. What's the best user API for this?**

Mitmproxy currently operates in a single mode: regular, reverse, transparent, upstream, or socks5. In all case, we spawn a single proxy server on port 8080 by default and listen for requests there. With upcoming features such as DNS support (#5232) and then QUIC, I want this to be more flexible.

#### 1. Improve Setup Experience

In an ideal world, I'd like users to be greeted with a setup screen that looks something like this:

> ###### mitmproxy is running!
>
> To get started, configure your client to use an HTTP proxy:
>  - ✅ HTTP Proxy running on port 8080.
>
> Alternatively, capture all traffic from this device or a specific app:
> - [Intercept this device's traffic button]
> - [Start Chrome instance button] 
> - [Start Firefox instance button]
> - [...]
>
> For advanced use cases, we've also set up the following services:
>  - ✅ SOCKSv5 Proxy running on port 1080.
>  - ✅ DNS Server running on port 5353.
>  - ✅ Transparent HTTP Proxy running on port 8081
>  - [Add reverse proxy button]
>  - [Add upstream proxy button]

#### 2. DNS/QUIC

For UDP-based protocols we absolutely need additional servers running anyways and not just one HTTP proxy. #5232 solves this by hardcoding a second DNS server, but that's not very extensible. While I strongly favor not planning for hypotheticals, I do think we could make good use of more extensibility here.

#### 3. OS Proxy Functionality

Ideally I want a simple "capture this device's traffic" toggle within mitmproxy (see above). This requires us to start and stop an additional (transparent) proxy on demand.

## Implementation Proposal

There are two main concerns: What does an ideal configuration language look like, and ideally how can we ensure backward compatibility? I'm generally willing to trade off backward compatibility for a nicer language, but maybe we can have both.

My current idea/proposal is to extend the `mode` option from `str` to a `list[str]` and extend the syntax from this:
```
proxy_mode [: mode_configuration]
```
to this:
```
proxy_mode [: mode_configuration] [@ [listen_addr:]listen_port]
```

The part after the `@` sign would then signal where we want to listen on (which is currently governed by listen_addr/listen_port). Some examples:

```shell
regular                          # a regular HTTP(S) proxy on port 8080
regular@8085                     # a regular HTTP(S) proxy on port 8085
regular@192.0.2.42:8080          # a regular HTTP proxy bound to an external IP
socks5                           # a SOCKSv5 proxy on port 8080
socks5@1080                      # a SOCKSv5 proxy on port 1080
reverse                          # invalid (required configuration missing)
reverse:https://example.com      # a reverse HTTP(S) proxy to example.com:443 on port 8080
reverse:https://example.com@443  # a reverse HTTP(S) proxy to example.com:443 on port 443
dns@53                           # a DNS server on port 53 (using the local resolver)
reverse-dns:8.8.8.8@53           # a DNS reverse proxy (forwarding all requests to 8.8.8.8)
```

This option could then default to something like `["regular", "socks5@1080", "dns@53"]` in line with the getting started screen above. If a user runs mitmproxy like so:
```
mitmproxy --mode transparent
```
Only a single transparent proxy on port 8080 would be started (keeping backward compatibility). Making use of the new feature, 
```
mitmproxy --mode regular --mode reverse:https://example.com@8085
```
would start a regular HTTP proxy and an additional reverse proxy.

--------------------------

@mitmproxy/devs and everyone else, what do you think? 😃

**Comment by meitinger** [comment]
While the proposed syntax offers a great balance between backwards compatibility and new functionality, there are two issues I'd like to raise with multiple-server mode:
1. mitmproxy supports a lot of options that are currently set globally, but should rather be set on a server-basis, e.g. `ignore-hosts`, `proxyauth` or `tcp-hosts` (later maybe also `udp-hosts` or `raw-hosts`). Of course one could make the argument, that settings those options individually is either beyond user-scope (and more addon-developer-scope), or that it will still require launching multiple mitmproxy instances. But with possible future HTTP/3<->HTTP/2 bridging and QUIC+DNS coexistence, it may not always be possible to separate protocols and conflicting global options.
2. A capture-all transparent proxy, which would be amazing, should also provide a more fine-graned filtering mechanism to the user. Here I'd suggest using the `[listen_addr:]listen_port` as filter, i.e. `[target_addr:]target_port`, since how mitmproxy itself communicates with this OS capture service should rather be hidden anyways. (On some OSes there may not even be a port used.)

So for the first point, how about doing something like:
```
regular-dns@53                                          # as dns@53 above, but keep the prefix for consistency
reverse[ignore-hosts=1.2.3.4]:https://example.com@8081  # only instance at port 8081 will ignore host 1.2.3.4
futureproto[param=value]@80;85-89                       # shorthand for six futureproto instances on port 80 and 85 thru 89,
                                                        # all with param set to value
reverse-dns:8.8.8.8@10.0.0.1:53;10.0.1.1:53             # two reverse DNS proxies, same upstream, listening on different IPs
regular[no-http2]@8080                                  # HTTP/1-only regular proxy on 8080
```
Of course instead of doing `--mode reverse-dns:8.8.8.8@53`, a new option could be introduced, e.g. `--instance dns[reverse=8.8.8.8]@53`. In this case `mode` is just another instance parameter.


For the second point, consider: 
`--mode reverse:https://example.com@8080 --mode oscapture:/dev/eth0 --mode regular-dns@53 --mode regular-quic@1-1024`
This could mean: Run a reverse proxy at local port 8080, then enable the OS capture for all the rest: If something is sent to port 53 on any remote target, use DNS. Otherwise use QUIC if the port number doesn't exceed 1024. 
`--mode oscapture[fallback]:/dev/eth0` might enable to use QUIC on port 53 still if DNS fails.

**Reply by mhils** [reply]
> mitmproxy supports a lot of options that are currently set globally, but should rather be set on a server-basis

Yes, good that you bring that up. For now I'm inclined to say that this is beyond scope. Of course in an ideal world it'd be nice to have "overlay configurations", but we should first see how far we can get without. It may not be worth the complexity in the end. It's much nicer for addon developers if there's a single global options state.

`regular-dns` instead of `dns` sounds good to me. 👍  Or we are very fancy and rename "regular" to "http" (regular can stay a deprecated alias for compat reasons)
Port ranges / multiple interfaces sound like a reasonable future extension, but let's keep things simple initially.

> For the second point, consider:
> `--mode reverse:https://example.com@8080 --mode oscapture:/dev/eth0 --mode regular-dns@53 --mode regular-quic@1-1024`
> This could mean: Run a reverse proxy at local port 8080, then enable the OS capture for all the rest: If something is sent to port 53 on any remote target, use DNS. Otherwise use QUIC if the port number doesn't exceed 1024.
> `--mode oscapture[fallback]:/dev/eth0` might enable to use QUIC on port 53 still if DNS fails.

Configurable/filterable OS proxying is a good idea, but let's not go down the rabbit hole where proxy modes start to influence each other or the order in which you specify things becomes important. OS Proxy mode should really just be transparent mode + automated redirection. Which of course means we somehow need transparent mode for UDP, but I think we can figure that one out. A hardcoded "53 = DNS, 443 = QUIC, Rest = Raw UDP" shall be good enough to start things off, in a second step we can then do packet inspection to figure out the protocol. :)

**Comment by Prinzhorn** [comment]
I have no objections for the suggested syntax. I'm mostly interested in how this can work in mitmdump (and also inside the mitmproxy UI) with dynamically changing the configuration (`running` event, error handling, can 3 of 4 listeners start and one crash? Can I add a new listener without affecting the others at all?). Also with all the events (e.g. `request`) what new properties does `Flow` need so we know where it came from? This is probably not part of the first iteration but maybe we need to talk about it to make sure we're not missing anything. But this should happen in a separate discussion.

**Reply by mhils** [reply]
> dynamically changing the configuration

We should do something similar to what we do for scripts: detect which ones are new, which ones have been removed, and only start/stop these. 

> can 3 of 4 listeners start and one crash?

On startup, I think we should fail if at least one port cannot be bound. Which probably means that we should take dns out of `auto` as binding to 53 is prone to permission errors on Linux. When mitmproxy is already running, I think we just log on startup errors and indicate it in the UI. I'm making all this up of course, if you feel that's ugly please shoot.

 >what new properties does `Flow` need so we know where it came from?

None. If you for some weird reason want different logic here, you can look at `flow.client_conn.sockname`. The most common use case I can think of involves multiple reverse proxies, and there you can do it by URL again.

**Reply by Prinzhorn** [reply]
> We should do something similar to what we do for scripts: detect which ones are new, which ones have been removed, and only start/stop these.

:+1: , would be great if there could be new hooks triggered either for each listener (start, stop, error) or one hook that tells you the listener configuration has changed (maybe the arguments tell you what changed).

> On startup, I think we should fail if at least one port cannot be bound. Which probably means that we should take dns out of `auto` as binding to 53 is prone to permission errors on Linux. When mitmproxy is already running, I think we just log on startup errors and indicate it in the UI. I'm making all this up of course, if you feel that's ugly please shoot.

I think that's a sensible default and if you use a single listener it's the same as the current behavior. However, I personally would like to fully remote control mitmdump, so being able to start it _without any listeners_ would be cool. So correct me if I'm wrong, we just need to make sure that `--no-server` will still work as expected but can dynamically be changed? I want "mitmdump process exited" and "could not listen" to be two separate responsibilities on my end. So I would always first start `mitmdump --no-server` and then configure it dynamically. But for now I just need single listener and kill the process if I need to change it. But once I have multiple listeners I don't want all of them to go down just because you changed one of the others.

> If you for some weird reason want different logic here

I was mostly interested in mitmproxy and mitmweb being able to display a column with info where the flow came from. Once we have multiple listeners this can be very relevant information. Even if we don't want another column (at least in mitmproxy for space reasons) the flow details page should show what listener configuration this flow went through. With the current single-listener process this information is implicit, with multiple listeners we need to be able to tell the user. We can't just dump all of them in the same bucket without being able to trace them back.

In Burp I was using 8080 for browser traffic and 8081 for mobile traffic. So I could easily filter them and have a clean separation within the same project file. That's what I personally want multiple listeners for. Plus a third listener for raw tcp maybe.

**Comment by mhils** [comment]
See https://github.com/mitmproxy/mitmproxy/pull/5393 for an initial prototype. :)

**Comment by ericbeland** [comment]
The command line passing and syntax seems like it might not scale that well to other capabilities--seems like it crams a lot into the command line with a new syntax to learn. The syntax isn't bad, though!

One alternative,  inspired by the BrowserMob proxy, is that there could be a two-stage start. First, you launch a "control plane" process that doesn't proxy anything, it just controls proxy-instances. 

In mitmweb, this would map roughly to the UI being described above by @mhils, but without any proxy-instances running.  In mitmweb, you could hit some buttons to launch proxy instances.

In mitmdump, this would map to a running background control-plane process that also would not have any "proxy instances" you could talk to _yet_, but it listens on a port. 

For mitmdump, from the command line, you could make subsequent command-line calls, one by one, that start  proxies, but use virtually the same command-line syntax that is used to launch _today's_ proxy. The "proxy-instance launching command," under the hood, would forward a **request-to-start** to the control plane process via the port.  The control-plane process would then start a _proxy-instance_ on a port, while also keeping track of success, failures, and remembering them for later cleanup.

There might be concerns I'm not considering, but thought I'd toss the idea out there.

**Reply by mhils** [reply]
Thanks for chiming in @ericbeland! :)

> The command line passing and syntax seems like it might not scale that well to other capabilities--seems like it crams a lot into the command line with a new syntax to learn.

To be clear, the proposed syntax is not only for command line invocations, but also what you would use in mitmproxy's `config.yaml`. I think your point on "new syntax to learn" is very reasonable though, if we need to add new syntax we should strive to keep it as simple as possible. Ideally we have a shiny UI in mitmproxy/mitmweb so that most users don't need to learn any syntax at all.

> In mitmweb, this would map roughly to the UI being described above by @mhils, but without any proxy-instances running. In mitmweb, you could hit some buttons to launch proxy instances.
>
> In mitmdump, this would map to a running background control-plane process that also would not have any "proxy instances" you could talk to _yet_, but it listens on a port.

I think we should definitely keep backwards compatibility with listening on 8080 by default (it's also a very reasonable thing to do). I already mentioned in #5393 that we should probably leave it at that as well if no other options are specified. 
mitmdump is *non-interactive* and wants to be used with a single invocation that can be easily killed. That being said, I absolutely don't object to an addon that exposes an API for control. :)

**Comment by ericbeland** [comment]
Ah, I see, I thought this was normally going to come in via the command line. I see your point re:interactivity.

**Comment by PavelICS** [comment]
Wouldn't it be too complex to have one mitm instance handling all servers?
I don't think it's a good idea combine mods like "regular", "upstream", "reverse" all together.

Maybe it would be better to create multiple instances of mitm Master in different threads. And group them together by some kind of control plane, something like MastersManager. In that case we add a new layer of MastersManager but internal structure of every Master by itself won't change a lot.
 
 Also like that we could dynamically add and stop Masters.

 It's very similar to the proposal by@ericbeland but in multi-threading way instead of multi-processing.

**Reply by mhils** [reply]
> I don't think it's a good idea combine mods like "regular", "upstream", "reverse" all together.

There are multiple drivers behind the multi-server mode. For example, I'd like to have a "capture all OS traffic" checkbox in mitmweb. This technically starts a new transparent proxy, i.e. an additional mode. Spawning both a regular and an upstream proxy at the same time also becomes possible, but that's more of an artifact. One could also solve this entire problem by hardcoding each special mode, but then we end up with duplicate settings like `listen_port`, `dns_listen_port`, `quic_listen_port` etc. I'd like to avoid that.

> Maybe it would be better to create multiple instances of mitm Master in different threads.

From my experience, adding multithreading or multiprocessing would introduce _much_ more complexity and subtle bugs. If you look at https://github.com/mitmproxy/mitmproxy/pull/5393, we don't need to change so much structurally - it's mostly the proxyserver addon that gets additional complexity. :)

**Comment by rbdixon** [comment]
This is a great feature and my top "missed" feature from other proxies I use frequently. I would want a way, from the API, to query which proxy port received the client connection. My usage for this is that I'd want to write extensions that treated traffic differently depending on which proxy port captured the request. I've done this most typically for request content decryption/encryption purposes where mitmproxy decrypts data, another tool mangles it, then mitmproxy re-encrypts the data. Right now I just use two mitmproxy instances.

**Reply by Prinzhorn** [reply]
> `Client.proxy_mode` is a thing now

https://github.com/mitmproxy/mitmproxy/pull/5393

I think that's what you mean? Because I had the same requirement to know where a flow/connection came from, see https://github.com/mitmproxy/mitmproxy/discussions/5288#discussioncomment-2833093

**Reply by rbdixon** [reply]
I see... yes this is all I needed:

```py
def request(flow):
    port = flow.client_conn.sockname[1]
    ctx.log.info(f'Proxy = {port}')
```

**Comment by mhils** [comment]
#5393 has been completed, thank you all for the fantastic feedback! :cake:

API users can use https://docs.mitmproxy.org/dev/api/mitmproxy/connection.html#Client.proxy_mode to distinguish where they are receiving requests from, I'd be happy to take PRs that add it to the UI as well.
